src/EPR.RegulatorService.Frontend.Web/Controllers/ReprocessorExporter/Registrations/ManageRegistrationsController.cs

```
@@ -1,26 +1,69 @@
 using EPR.RegulatorService.Frontend.Web.ViewModels.ReprocessorExporter;
+
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
 using Microsoft.FeatureManagement.Mvc;
 
-namespace EPR.RegulatorService.Frontend.Web.Controllers.ReprocessorExporter.Registration;
+namespace EPR.RegulatorService.Frontend.Web.Controllers.ReprocessorExporter.Registrations;
 
 [FeatureGate(FeatureFlags.ReprocessorExporter)]
 [Route($"{PagePath.ReprocessorExporterRegistrations}/{PagePath.ManageRegistrations}")]
 public class ManageRegistrationsController : Controller
 {
-    [HttpGet]
-    public IActionResult Index()
+    private readonly IRegistrationService _registrationService;
+    private readonly ILogger<ManageRegistrationsController> _logger;
+
+    public ManageRegistrationsController(IRegistrationService registrationService, ILogger<ManageRegistrationsController> logger)
     {
-        ViewBag.BackLinkToDisplay = "";
+        _registrationService = registrationService;
+        _logger = logger;
+    }
 
-        var model = new ManageRegistrationsViewModel
+    [HttpGet]
+    [ProducesResponseType(StatusCodes.Status200OK)]  // When the view loads successfully
+    [ProducesResponseType(StatusCodes.Status302Found)] // When redirecting to the error page
+    public IActionResult Index([FromQuery] int id)
+    {
+        try
         {
-            ApplicationOrganisationType = ApplicationOrganisationType.Reprocessor
-        };
+            // Validate ID
+            if (id <= 0)
+            {
+                _logger.LogError("Invalid ID received in query string: {Id}", id);
+                return RedirectToAction(PagePath.Error, "Error");
+            }
+
+            var registration = _registrationService.GetRegistrationById(id);
+
+            if (registration == null)
+            {
+                _logger.LogError("No registration found for ID: {Id}", id);
+                return RedirectToAction(PagePath.Error, "Error");
+            }
```

My feeling is that consumers of the service shouldn't perform null checking. This is a Get method not a TryGet, so the expectation is that something will be returned. If the service threw a custom exception if the item wasn't found, then you wouldn't need this check and nor would other consumers of the service method. The try/catch could then handle the exception logging for you and the redirect.

---

Agreed, I will remove the check within the controller. I will assume at this point in time that the real version of GetRegistrationById will throw the exception, which will then be handled within the controller.

---

What do you think about introducing something in the mock so that we could test it? For example if someone supplies the id 99999, we could throw an exception just on that id?

---

....that would also allow you to write the customer exception and ensure that any logging works correctly.

---

I will add the 99999 condition and add a test to the mocked service. I am thinking the only way to test that in the controller is to mock the middleware to handle the exception? Unless you have another idea?